platform/linux-generic/odp_ipsec.c

```diff
@@ -596,6 +598,42 @@ static int ipsec_in_ah_post(odp_packet_t pkt,
 	return 0;
 }
 
+static void
+ipsec_sa_err_stats_update(odp_ipsec_packet_result_t *result)
```

Comment:
Now that the call sites of this changed, this could take SA (as ipsec_sa_t) and status (odp_ipsec_op_status_t) as parameters and get rid of the _odp_ipsec_sa_entry_from_hdl() call and all the result dereferencing.

Is there some benefit in calling this from the top level API functions instead of ipsec_in_single() and ipsec_out_single()? That puts the call in 6 call sites instead of 2.

Reply:
> Now that the call sites of this changed, this could take SA (as ipsec_sa_t) and status (odp_ipsec_op_status_t) as parameters and get rid of the _odp_ipsec_sa_entry_from_hdl() call and all the result dereferencing.

Agreed.
 
> Is there some benefit in calling this from the top level API functions instead of ipsec_in_single() and ipsec_out_single()? That puts the call in 6 call sites instead of 2.

In odp_ipsec_out_inline(), there are errors set after ipsec_out_single() call. For inbound, we can put the call in ipsec_in_single().

Reply:
I think the call to ipsec_sa_err_stats_update() in odp_ipsec:_out_inline() should be made a bit earlier. It should be in the preceding else branch before the error result queuing. That way error counters get updated even if the application lets the result queue sometimes become full. The call also otherwise fits better there since we do not need to update error stats for non-errors.

Reply:
Agreed.